### ISSUE TITLE
Send multiple allowance transactions parameters and return data

### ISSUE THREAD

**Issue by JanellC** [US-EPA-CAMD/PI-1#1258]
### Definition of Ready
- [x] Does this meet the INVEST criteria?
- [x] Does this have a clearly defined user?
- [x] Does this have an estimate?
- [x] Does this have acceptance criteria?
- [x] Does this have sufficient context (wireframes, business rules, etc.)?
- [x] Is this traceable to a feature, epic, or PI?

## User Story
As a CAMPD API user I want to send filter criteria to the  allowance transactions API endpoint so that I can see results

## Acceptance Criteria
Given that a user has entered one or more valid values for all the parameters (see parameters below)
When API endpoint is executed
Then the system displays the records that match the criteria (containing data elements below)

Given that a user has entered one or more valid values for all the parameters in an upper case or lower case format
When API endpoint is executed
Then the system displays the records that match the criteria

Given that a user has entered one or more valid values for all the parameters without white space(s) before or after the delimiter
When API endpoint is executed
Then the system displays the records that match the criteria

Given that a user has entered one or more valid values for all the parameters with white space(s) before or after the delimiter
When API endpoint is executed
Then the system displays the records that match the criteria

Given that a user has made a GET request with the attach file parameter set to true
When request is executed
Then the system produces a file in the requested file format (json or csv)
AND the file contains the records that match the criteria (see context)

### Does this Story require 508 Testing?
- [ ] Implementing a new authoring tool that will produce code / content for the application
- [ ] Implementing a 3rd party widget that will be part of the application
- [ ] Adding new / updating custom functionality 
- [ ] Adding new / updating a page template
- [ ] Adding new / replacing existing functionality (unless the replacement has already passed 508 testing)
- [ ] Full Site Review prior to a scheduled EPA 508 review
- [ ] Planning a video that will appear on the application
- [ ] Producing a video that will appear on the application
- [ ] Producing Training materials
- [ ] Producing documents

## Parameters 

- Program: Optional, multiple values allowed, retired/active allowance programs, values: ARP, CAIRNOX, CAIROS, CAIRSO2, CSNOX, CSNOXOS, CSOSG1/2, CSSO2G1/2, NBP, OTC, TXSO2

- Account Type: Optional, multiple values allowed, values: Facility Account, Unit Account, General Account, Reserve Account, Surrender Account, AL (UI only)L, searches both buyer and seller account types
 
- Account Number: Optional, multiple values allowed, searches both buyer and seller Account numbers
 
- Facility ID (ORIS code): Optional, multiple values allowed, searches both buyer and seller ORIS codes
  
- Owner/ Operator:  Optional, multiple values allowed, query will return all records that match owner OR operator specified, searches both buyer and seller Owner/Op
 
- State:  Optional, multiple values allowed, 2 letter postal abbreviation, searches both buyer and seller states
 
- Vintage Year: Optional, multiple values allowed, Format: YYYY, 1995 or greater 
 
- Transaction Begin Date: Required, single value,  Range: 03/23/1993-current date Format: YYYY-DD-MM
 
- Transaction End Date: Required, single value, Range: 03/23/1993-current date, Format: YYYY-DD-MM
 
- Transaction Type: Optional, multiple values allowed, Value: Initial Allocation, Emissions Deduction, Private Transfer, New Unit Aside Allocation, etc.  

## Data Elements 
When the system identifies at least one record that matches the criteria, then for each record the system will display the data below and ordered by PRG_CODE> TRANSACTION_ID > VINTAGE_YEAR > START_BLOCK

PRG_CODE
TRANSACTION_ID
TRANSACTION_TOTAL
TRANSACTION_TYPE
SELL_ACCT_NUMBER
SELL_ACCT_NAME
SELL_ACCOUNT_TYPE
SELL_FACILITY_NAME
SELL_ORISPL_CODE
SELL_STATE
SELL_EPA_REGION
SELL_SOURCE_CAT
SELL_OWN_DISPLAY_NAME
BUY_ACCT_NUMBER
BUY_ACCT_NAME
BUY_ACCOUNT_TYPE
BUY_FACILITY_NAME
BUY_ORISPL_CODE
BUY_STATE
BUY_EPA_REGION
BUY_SOURCE_CAT
BUY_OWN_DISPLAY_NAME
TRANSACTION_DATE
VINTAGE_YEAR
START_BLOCK
END_BLOCK
TOTAL_BLOCK

## Context
- need regression testing for allowance holdings 

[CAMPD_DataLables_Allowance_5_12_2021.xlsx](https://teams.microsoft.com/l/file/C57F5045-F938-472B-AE6E-51BF711FFA12?tenantId=88b378b3-6748-4867-acf9-76aacbeca6a7&fileType=xlsx&objectUrl=https%3A%2F%2Fusepa.sharepoint.com%2Fsites%2FCAMDCVPTeam%2FShared%20Documents%2FGeneral%2FCoding%20%26%20Error%20Standards%2FCAMPD_DataLables_Allowance_5_12_2021.xlsx&baseUrl=https%3A%2F%2Fusepa.sharepoint.com%2Fsites%2FCAMDCVPTeam&serviceName=teams&threadId=19:6f05dfa026c84672a2cc1bfa72c58f14@thread.skype&groupId=21a5e1cd-d3f1-48a6-801f-0d34e2d63d23)

Travis,

CAMDNATS.TTRANTYPE lists the possible transaction types for transactions recorded in the CAMDNATS.TTRANSACT table. CAMDNATS only includes legacy OTC and NBP data.

CAMDAMS.TRANSACTION_TYPE_CODE lists the possible transaction types for transactions recorded in the CAMDAMS.TRANSACTION table. This covers all transactions for all other programs (ARP, CAIR, CSAPR, any new program that might come along).

ALLOCATION_IND in CAMDAMS.TRANSACTION_TYPE_CODE has 3 possible values.
• 0 – transaction type that is not an allocation
• 1 – transaction type that is an allocation
• -1 – transaction type that acts as a reversal or surrender of an allocation
o These are exclusively ARP transaction types that were used during ARP Phase 1
 DI Reallocation Surrender
 RP Phase 2 Early Reduction Payback
 RS Reduced Utilization Payback
 SP State Cap Payback
 SS Substitution Payback
 ST Substitution Termination
 TM Reduced Utilization Termination

The ALLOCATION_IND field is used during the AMPD data mart loading on CAMDPUB when populating the “Compliance Year Allowances Allocated” for compliance query results.

It is also used during the running of compliance in CBS and CSA when determining which allowances to deduct using FIFO.

Mark

From: Lageman, Travis Lageman.Travis@epa.gov
Sent: Monday, April 12, 2021 2:03 PM
To: Mark Hayward Mark.Hayward@erg.com; Ann Albrecht Ann.Albrecht@erg.com
Cc: Desantis, Laurel Desantis.Laurel@epa.gov; Kuhns, Jason Kuhns.Jason@epa.gov; Stetsyuk, Karina Stetsyuk.Karina@epa.gov; Gill, Sandeep Gill.Sandeep@epa.gov
Subject: Lookup code table for Allowance Transactions

CAUTION: Don't open links or attachments unless you recognize the sender and know they are safe.
Hi Mark & Ann,

Hope you had a nice weekend. I have another database question. I was looking for the transaction type code table and realized there are two different tables, one under the TRANSACTION_TYPE_CODE in CAMDAMS (which I believe is the current) and then one under TTRANTYPE in CAMDNATS (the older). I believe these are both necessary to get the full list of transaction types in the database. Is this correct or is there additional context I should be aware of?

I also noticed that there are some -1 values in the ALLOCATION_IND field for TRANSACTION_TYPE_CODE which I am confused as to what that identifies as opposed to 0 or 1, which make sense.

Thanks in advance,

Travis Lageman
Environmental Policy Analyst
Clean Air Markets Division
U.S. Environmental Protection Agency
P: 202-564-0520

## Testing
Test Case: [TC311 ](https://cvpcorp.testquality.com/project/11428/plan/21550/test/296747)
Tester: @JanellC 

### Definition of Done
- [x] Have all Acceptance Criteria been agreed to and validated by the scrum team?
- [x] Has the code been unit tested and peer reviewed?
- [x] Have the functional tests been executed?
- [x] Have all defects been reviewed and dispositioned, resolved, or deferred?
- [ ] Has the user story been accepted by the Product Owner?
- [x] Is the user story ready to be deployed to the test/staging environments?

**Comment by tlageman-epa** [comment]
@JanellC @sdhawan16  I noticed that a number of the json property labels returning via swagger for transactions do not match the labels in the data dictionary. I thought we were using those agreed upon labels for new data types so we don't have to go back and update these later?

**Comment by JanellC** [comment]
Hi @tlageman-epa  during backlog refinement on 8/11, the dev team mentioned that they reused some of the code from allowance holdings to produce the allowance transactions labels. Updating the JSON in this ticket might have broken some stuff since we use the same labels for holdings and transactions. We have zenhub ticket Update account data labels to the new data dictionary #1904 in the backlog to update the data labels for allowance holdings and transactions

**Comment by JanellC** [comment]
TC311 passed @tlageman-epa 

https://easey-dev.app.cloud.gov/api/account-mgmt/allowance-transactions?page=1&perPage=10&transactionBeginDate=2000-03-01&transactionEndDate=2000-03-01



![image.png](https://images.zenhubusercontent.com/5f5928c1045cffdce7ce89c3/99931c98-1058-44fd-993e-061cdbb10d3e)
